### ISSUE TITLE
Issue parsing Edmx version in temp metadata file

### ISSUE THREAD

**Issue by francisharvey** [OData/odata.net#1475]
Hi guys,

Is it possible something went wrong in last few release?
It may perhaps be linked with the Edmx 4.01 support or modifications done after v7.5.1?

As someone pointing out here: "https://github.com/microsoft/Dynamics-AX-Integration/issues/88",
I'm also having an issue instanciating the Resources class, which was working fine for months,
with a dll I had previously generated with the version 7.5.1...

For some other reason I had to update all projects target framework to .NET 4.7 and I then updated all nuget to latest version. I also regenerated the client code with success.

Now I'm unable to instanciate the resources class and the exception thrown talk about having a "4" somewhere at the begining of a name.
I even tried using the OData v7.6-beta but I got the same issue.

Since my client code is generated dynamically, I don't have access to the underlying code in my application... but I'm able to see few things in another test app with the full generated code.

In the Resources contructor line 31, "This.Format.UseJson" is where it fails.
The exceptions: 
System.TypeInitializationException: 'The type initializer for 'GeneratedEdmModel' threw an exception.'
InvalidOperationException: Name cannot begin with the '4' character, hexadecimal value 0x34. Line 1, position 22.; 

I also found as @rajdil says, the line 1 and position 22 refer to the Version attribute in the Edmx temp metadata file.

So is it somehow related to .NET4.7 ?, or a bug in the client code generated?
Is there some people that are using .NET4.7 with OData?

Any suggestion on what I can do to fix this?

Thanks, 
Francis Harvey

**Comment by francisharvey** [comment]
This method is the one that throws errors: global::Microsoft.OData.Edm.Csdl.CsdlReader.TryParse(reader, true, out edmModel, out errors)

Getting rid of the Prefix, version attribute and everything else referencing them in the metadata file solve the issue, but clearly not a viable thing to ask to a customer.

Here's what now look like the tempsmetadata file: 

`<Edmx> <DataServices>   <Schema Namespace=""Microsoft.Dynamics.DataEntities"" xmlns=""http://docs.oasis-open.org/odata/ns/edm"">`

and what it was before:

`<edmx:Edmx Version=""4.0"" xmlns:edmx=""http://docs.oasis-open.org/odata/ns/edmx"">
  <edmx:DataServices>
    <Schema Namespace=""Microsoft.Dynamics.DataEntities"" xmlns=""http://docs.oasis-open.org/odata/ns/edm"">`

**Comment by francisharvey** [comment]
Hi @mikepizzo, 

Do you think you latest changes done for https://github.com/OData/odata.net/issues/1470 is related or may fix this issue?

Thanks
Francis Harvey

**Comment by mikepizzo** [comment]
That was my hope in fixing it, but from your expanded description I don't see that it's related.

If I understand your comments above, the second sample (with the Version and other Edmx attributes) was not working, but the one where you removed the Version attribute and namespace does work?

The text in the examples above is incomplete, so I modified it by closing all of the tags and the following code worked fine:

~~~c#
public void ValidateEdmxVersions(string odataVersion)
{
  string xml = @"<edmx:Edmx Version=""4.0"" xmlns:edmx=""http://docs.oasis-open.org/odata/ns/edmx""> <edmx:DataServices> <Schema Namespace=""Microsoft.Dynamics.DataEntities"" xmlns=""http://docs.oasis-open.org/odata/ns/edm""></Schema></edmx:DataServices></edmx:Edmx>";
  var stringReader = new System.IO.StringReader(xml);
  var xmlReader = System.Xml.XmlReader.Create(stringReader);

  IEdmModel edmModel = null;
  IEnumerable<EdmError> edmErrors;
  CsdlReader.TryParse(xmlReader, out edmModel, out edmErrors);
  if (edmModel != null)
  {
       Console.WriteLine(edmErrors.Count(), edmModel.GetEdmVersion());
  }
}

~~~

On the other hand, trying the version with the Version and namespace removed does not work, so maybe I'm misunderstanding your description of the issue.

*Can you send me a complete tempsmetadata file that repos the problem?, along with the full error description (you mention something about names not starting with "4"?)

**Comment by francisharvey** [comment]
Mike,

Unfortunately, it seems the workaround only fools the TryParse...
All my queries are now failing in DataServiceQuery(Of T).Execute() with :

System.ArgumentOutOfRangeException: 'Length cannot be less than zero.
Parameter name: length'
   at System.String.Substring(Int32 startIndex, Int32 length)
   at Microsoft.OData.TypeUtils.ParseQualifiedTypeName(String qualifiedTypeName, String& namespaceName, String& typeName, Boolean& isCollection)
   at Microsoft.OData.JsonLight.ODataJsonLightContextUriParser.ResolveType(String typeName, Func`3 clientCustomTypeResolver, Boolean throwIfMetadataConflict)
   at Microsoft.OData.JsonLight.ODataJsonLightContextUriParser.ParseContextUriFragment(String fragment, Func`3 clientCustomTypeResolver, Boolean throwIfMetadataConflict, Boolean& isUndeclared)
   at Microsoft.OData.JsonLight.ODataJsonLightContextUriParser.ParseContextUri(ODataPayloadKind expectedPayloadKind, Func`3 clientCustomTypeResolver, Boolean throwIfMetadataConflict)
   at Microsoft.OData.JsonLight.ODataJsonLightContextUriParser.Parse(IEdmModel model, String contextUriFromPayload, ODataPayloadKind payloadKind, Func`3 clientCustomTypeResolver, Boolean needParseFragment, Boolean throwIfMetadataConflict)
   at Microsoft.OData.JsonLight.ODataJsonLightDeserializer.ReadPayloadStart(ODataPayloadKind payloadKind, PropertyAndAnnotationCollector propertyAndAnnotationCollector, Boolean isReadingNestedPayload, Boolean allowEmptyPayload)
   at Microsoft.OData.JsonLight.ODataJsonLightPayloadKindDetectionDeserializer.DetectPayloadKind(ODataPayloadKindDetectionInfo detectionInfo)
   at Microsoft.OData.Json.ODataJsonFormat.DetectPayloadKindImplementation(ODataMessageInfo messageInfo, ODataMessageReaderSettings settings)
   at Microsoft.OData.ODataMessageReader.DetectPayloadKind()
   at Microsoft.OData.Client.Materialization.ODataMaterializer.CreateODataMessageReader(IODataResponseMessage responseMessage, ResponseInfo responseInfo, ODataPayloadKind& payloadKind)
   at Microsoft.OData.Client.Materialization.ODataMaterializer.CreateMaterializerForMessage(IODataResponseMessage responseMessage, ResponseInfo responseInfo, Type materializerType, QueryComponents queryComponents, ProjectionPlan plan, ODataPayloadKind payloadKind)
   at Microsoft.OData.Client.MaterializeAtom..ctor(ResponseInfo responseInfo, QueryComponents queryComponents, ProjectionPlan plan, IODataResponseMessage responseMessage, ODataPayloadKind payloadKind)
   at Microsoft.OData.Client.QueryResult.CreateMaterializer(ProjectionPlan plan, ODataPayloadKind payloadKind)
   at Microsoft.OData.Client.QueryResult.ProcessResult[TElement](ProjectionPlan plan)
   at Microsoft.OData.Client.DataServiceRequest.Execute[TElement](DataServiceContext context, QueryComponents queryComponents)
   at Microsoft.OData.Client.DataServiceQuery`1.GetEnumerator()
   at System.Linq.SystemCore_EnumerableDebugView`1.get_Items() 

Unless it's because i'm using DynamicLinq and my resulting elementType is 
"DynamicClass1, DynamicClasses, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null", but this was working very well before.

This is the metadata downloaded from the server (not the temp one), it has been modified to be 'valid'...
Yes, D365FO server metadata are not valid OOB. :-/ 
[metadata.zip](https://github.com/OData/odata.net/files/3253014/metadata.zip)

This is the temp one: 
[tempMetadata.zip](https://github.com/OData/odata.net/files/3253084/tempMetadata.zip)

In the mean time, I will also try to create a new project and give it another try in VS.

Thank you,
Francis Harvey

**Comment by francisharvey** [comment]
Hi @mikepizzo,

I confirm the same behavior occurs in VS2017 with the 7.5.1 client tool.
Updating nuget to 7.5.4 did not fix anything.

Regards,
Francis Harvey

**Comment by mikepizzo** [comment]
The issue with tempMetadata.xml is that all of the quotes are doubled, as if the string is intended to be embedded within another string, rather than the contents of a file. 

If I do a replace of "" with " within the file, and then save, everything seems to work fine.

Can you elaborate on the steps that you're going through to download and save the file?

**Comment by rajdil** [comment]
Hi Francis Harvey

You can get rid of the problem by removing the "" to " in the entire file.

Rest all should be fine.
Thanks
Raj

**Comment by francisharvey** [comment]
@mikepizzo, It is the client code that generate the temp file, I input the original metadata and it produces the temp file.

It also append the original metadata to the file rather than overwriting it, so you can end up with more than 60mb of metadata in text format, multiple <edmx:Edmx> tags pair in the same file.

By the way, i'm not sure it is really needed to use the temp file when the metadata URI is already a File://
Since you cannot use authentication with the generator, it will probably be a File:// most of the times.

In the first implementation of our connector, I had modified the tt file to produce an array of string representing every lines of the metadata with a string.split, and using a string.join at runtime to bring back the model in the variable. The original issue was because the metadata was too big to be in a string for the compiler, so splitting it in lines had solved the issue... 

With the version using the temp file, my solution was somehow equal to yours, so I did not bother to port my fix to the new tt version.

In short, I don't think I have control over the temp file generation... 
Now I dynamically use the ODataT4CodeGenerator.TransformText() and compile the result when needed.

It would also be great to have some kind of access to the generator, I think it is 'internal' at this time.

Thanks

**Comment by XavierAnnabi** [comment]
Hi guys,
Confirmed, am getting the same issue the last release has bug, usually when i generate my proxy class I get  errors related to namespace Microsoft.OData.Edm.Library but this time no errors, but am getting 
### **The type initializer for 'GeneratedEdmModel' threw an exception**, it worked before I generate with the last release.

Thanks,

**Comment by mikepizzo** [comment]
Can someone confirm the version where this worked and the first version where this failed? I'm trying to track down any change that might have caused this, but don't see any changes to that part of the code for quite a while.

**Comment by KanishManuja-MS** [comment]
@francisharvey Can you make sure that the temp file path is empty if it is a file in the client code generator? 
@mikepizzo I suspect it is because of the temp file I introduced for storing large metadata files in the client code generator.

**Comment by mikepizzo** [comment]
@KanishManuja-MS; sounds likely. Do you have a link to that commit? How easy is this to fix?

**Comment by francisharvey** [comment]
Hi @KanishManuja-MS, in fact, D365FO has a VERY large metadata file, so I need the temp file to be able to compile the dll. The first version on which I started the project was 7.5.0 in August 2018, but I had customized the TT file to produce the metadata string array to bypass the litteral string limit has I explained in the comment before Xavier's. Perhaps @XavierAnnabi has a later working version number in mind...?

**Comment by XavierAnnabi** [comment]
Hi,
am working with dynamics NAV 2017 and 2018 the issue is related to NAV 2017 Odata V4.
with NAV 2018 when i generate the proxyclass is works but with NAV 2017 id doesn't and am getting   
The type initializer for 'GeneratedEdmModel' threw an exception.

Thanks,

**Comment by francisharvey** [comment]
@XavierAnnabi, Mike was looking for a version/commit of the OData package to narrow down the issue.

I can say I downloaded source code on March 20th, and I took the ODataT4CodeGenerator class (with temp metadata file) and it was working fine since. So the issue does not seems to be directly related to @KanishManuja-MS temp file commit, but maybe few commits after...

The issue was discovered only few weeks ago when I changed the target framework and reinstall all nuget latest version.

**Comment by robertmclaws** [comment]
I ran into this same issue when trying to use the CsdlReader to try to parse an Entity Framework 6.2 EDMX file. Feels like the parser should be able to handle node namespaces. maybe it needs to use Linq to XML instead?

**Comment by xuzhg** [comment]
Fixed at c719a53dfc089bc8e2c191532bed25d6d28f917e

**Comment by Rickrat** [comment]
@francisharvey  I have a solution for you.

      internal static Version ReturnEdmxVersion(string fileName)
      {

         XDocument doc = XDocument.Load(fileName);
         var versionAttr = doc.Document.Root?.Attributes().FirstOrDefault(x => x.Name.LocalName == "Version");
         string value = versionAttr.Value;

         return Version.Parse(value);
      }

**Comment by francisharvey** [comment]
@Rickrat Thanks!

The issue with the temp metadata file has already been fixed for a while by @odero if i'm not mistaken.
Thank everyone, great teamwork finding and fixing it!

Francis Harvey
